File: agent/sdk/python/examples/service_lifecycle.py
from dataclasses import dataclass
from typing import Literal

from layerx_sdk import Client, IdempotentMutation, SubmissionState


@dataclass(frozen=True)
class ServiceActivity:
    stage: Literal["commit", "deliver", "accept"]
    canonical_bytes: bytes


def service_lifecycle(
    client: Client,
    activities: list[IdempotentMutation[ServiceActivity]],
) -> list[SubmissionState]:
    observed: list[SubmissionState] = []
    for activity in activities:
        result = client.call("submit", activity)
        if not isinstance(result, SubmissionState.__args__):
            raise TypeError("invalid_submission_state")
        observed.append(result)
    return observed

File: agent/sdk/typescript/examples/service-lifecycle.ts
import { Client, type IdempotentMutation, type SubmissionState } from "../src/index.js";

export interface ServiceActivity { readonly stage: "commit" | "deliver" | "accept"; readonly canonicalBytes: Uint8Array }

export async function serviceLifecycle(
  client: Client,
  activities: readonly IdempotentMutation<ServiceActivity>[],
): Promise<SubmissionState[]> {
  const observed: SubmissionState[] = [];
  for (const activity of activities) {
    observed.push(await client.call("submit", activity));
  }
  return observed;
}
